File: Nsigii_EchoRelayServer/nsiggi.py
import secrets
import time
import math
from enum import Enum

class State(Enum):
    YES = 1
    NO = 0
    MAYBE = -1

class NSIGIIMessage:
    def __init__(self, data: list[State]):
        self.data = data
        self.entropy_bits = []  # Measured entropy for each state
    
    def measure_channel_entropy(self) -> float:
        """Measure real entropy from system noise sources"""
        # Use secrets for true entropy, not random
        noise_samples = [secrets.randbits(8) for _ in range(len(self.data))]
        
        # Calculate Shannon entropy of the noise floor
        entropy = 0.0
        for sample in noise_samples:
            p = sample / 255.0
            if p > 0:
                entropy -= p * math.log2(p)
        
        return entropy  # Returns bits of entropy per symbol
    
    def encode_sparse(self):
        """Create sparse matrix with θ states preserved"""
        # For MAYBE states, we embed measured entropy
        matrix = []
        for i, state in enumerate(self.data):
            if state == State.MAYBE:
                # θ gets its own entropy vector
                theta_entropy = self.measure_channel_entropy()
                matrix.append([1, 0, 1, 0])  # θ pattern [1,0,1,0]
                self.entropy_bits.append(theta_entropy)
            elif state == State.YES:
                matrix.append([1, 1, 0, 0])  # YES pattern
                self.entropy_bits.append(0.0)
            else:  # NO
                matrix.append([0, 0, 1, 1])  # NO pattern
                self.entropy_bits.append(0.0)
        return matrix

class NSIGIIEchoServer:
    def __init__(self):
        self.force_newtons = 1.25  # From your physics model
        self.distance_meters = 15.0
        self.theta_angle = math.radians(30)  # cos(θ) = 0.866
        
    def calculate_work(self, message: NSIGIIMessage) -> float:
        """Calculate work = F × d × cos(θ)"""
        # Add entropy cost for MAYBE states
        base_work = self.force_newtons * self.distance_meters * math.cos(self.theta_angle)
        entropy_cost = sum(message.entropy_bits) * 0.1  # Scale factor
        return base_work + entropy_cost
    
    def apply_isomorphism(self, message: NSIGIIMessage):
        """Create conjugate pair: YES↔NO, MAYBE preserved"""
        transformed = []
        for state in message.data:
            if state == State.YES:
                transformed.append(State.NO)
            elif state == State.NO:
                transformed.append(State.YES)
            else:  # MAYBE is sacred
                transformed.append(State.MAYBE)
        return NSIGIIMessage(transformed)
    
    def verify_echo(self, sent: NSIGIIMessage, echoed: NSIGIIMessage) -> bool:
        """Verify via XOR: sent ⊕ echoed → ε or θ"""
        if len(sent.data) != len(echoed.data):
            return False
        
        for s, e in zip(sent.data, echoed.data):
            # XOR operation for trinary states
            if s == State.MAYBE and e == State.MAYBE:
                # θ ⊕ θ = 0 (verified as MAYBE)
                continue
            elif (s == State.YES and e == State.NO) or \
                 (s == State.NO and e == State.YES):
                # Conjugate pair verified
                continue
            else:
                # Invalid state transition
                return False
        
        # Check work calculation matches
        sent_work = self.calculate_work(sent)
        echoed_work = self.calculate_work(echoed)
        
        # Work difference must be within θ tolerance
        work_delta = abs(sent_work - echoed_work)
        return work_delta < 0.5  # 0.5 Joule tolerance

# Demo: "I Built a Consensus Protocol That Never Ghosts You"
if __name__ == "__main__":
    # Create message with MAYBE states
    message = NSIGIIMessage([
        State.YES, State.MAYBE, State.NO, State.MAYBE
    ])
    
    server = NSIGIIEchoServer()
    
    print("=== NSIGII Echo Verification ===")
    print(f"Original: {[s.name for s in message.data]}")
    
    # Measure entropy (this is the θ measurement)
    entropy = message.measure_channel_entropy()
    print(f"Channel Entropy: {entropy:.4f} bits/symbol")
    
    # Calculate work required
    work = server.calculate_work(message)
    print(f"Work Required: {work:.4f} Joules")
    
    # Apply isomorphism (what server echoes back)
    echo = server.apply_isomorphism(message)
    print(f"Echoed: {[s.name for s in echo.data]}")
    
    # Verify
    is_verified = server.verify_echo(message, echo)
    print(f"Verification: {'✓ VERIFIED' if is_verified else '✗ FAILED'}")
    
    # Simulate bit rate bounds
    bitrate = len(message.data) / (work / 10)  # Simplified
    print(f"Effective Bit Rate: {bitrate:.2f} messages/Joule")

File: Nsigii_EchoRelayServer/nsiggi.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/random.h>
#include <math.h>
#include <string.h>

// θ state as 2-bit entropy register
typedef enum {
    STATE_NO = 0b00,      // 0
    STATE_YES = 0b01,     // 1
    STATE_MAYBE = 0b10,   // θ (entropy residue)
    STATE_EPSILON = 0b11  // Channel clear
} trinary_state_t;

typedef struct {
    trinary_state_t state;
    double entropy;  // Measured entropy for this symbol
} nsigii_symbol_t;

typedef struct {
    nsigii_symbol_t *symbols;
    size_t length;
    double *entropy_bits;  // Per-symbol entropy measurements
} nsigii_message_t;

// Measure REAL entropy from OS, not pseudorandom
double measure_channel_entropy(size_t num_samples) {
    unsigned char buffer[64];
    double entropy = 0.0;
    
    // Use getrandom() for cryptographic entropy
    if (getrandom(buffer, sizeof(buffer), 0) != sizeof(buffer)) {
        perror("getrandom");
        return -1.0;
    }
    
    // Calculate Shannon entropy of the noise floor
    int freq[256] = {0};
    for (int i = 0; i < 64; i++) {
        freq[buffer[i]]++;
    }
    
    for (int i = 0; i < 256; i++) {
        if (freq[i] > 0) {
            double p = (double)freq[i] / 64.0;
            entropy -= p * log2(p);
        }
    }
    
    return entropy;
}

// Create sparse matrix encoding with θ preservation
void encode_sparse(nsigii_message_t *msg) {
    msg->entropy_bits = malloc(msg->length * sizeof(double));
    
    for (size_t i = 0; i < msg->length; i++) {
        if (msg->symbols[i].state == STATE_MAYBE) {
            // θ gets its own entropy measurement
            msg->entropy_bits[i] = measure_channel_entropy(64);
            msg->symbols[i].entropy = msg->entropy_bits[i];
            // Sparse pattern for θ: [1,0,1,0]
            printf("θ[%zu] entropy: %.4f bits\n", i, msg->entropy_bits[i]);
        } else if (msg->symbols[i].state == STATE_YES) {
            msg->entropy_bits[i] = 0.0;
            // YES pattern: [1,1,0,0]
        } else {  // STATE_NO
            msg->entropy_bits[i] = 0.0;
            // NO pattern: [0,0,1,1]
        }
    }
}

// Calculate work = F × d × cos(θ)
double calculate_work(nsigii_message_t *msg, double force, double distance) {
    double base_work = force * distance * cos(30.0 * M_PI / 180.0); // cos(30°) = 0.866
    
    // Add entropy cost for MAYBE states
    double entropy_cost = 0.0;
    for (size_t i = 0; i < msg->length; i++) {
        entropy_cost += msg->entropy_bits[i] * 0.1;  // Scale factor
    }
    
    return base_work + entropy_cost;
}

// Apply isomorphic transformation (YES↔NO, MAYBE preserved)
void apply_isomorphism(nsigii_message_t *src, nsigii_message_t *dst) {
    dst->length = src->length;
    dst->symbols = malloc(dst->length * sizeof(nsigii_symbol_t));
    dst->entropy_bits = malloc(dst->length * sizeof(double));
    
    for (size_t i = 0; i < src->length; i++) {
        switch (src->symbols[i].state) {
            case STATE_YES:
                dst->symbols[i].state = STATE_NO;
                break;
            case STATE_NO:
                dst->symbols[i].state = STATE_YES;
                break;
            case STATE_MAYBE:
                dst->symbols[i].state = STATE_MAYBE;  // Sacred
                break;
            default:
                dst->symbols[i].state = STATE_EPSILON;
        }
        dst->symbols[i].entropy = src->symbols[i].entropy;
        dst->entropy_bits[i] = src->entropy_bits[i];
    }
}

// XOR verification: sent ⊕ echoed → ε or θ
int verify_echo(nsigii_message_t *sent, nsigii_message_t *echoed) {
    if (sent->length != echoed->length) return 0;
    
    for (size_t i = 0; i < sent->length; i++) {
        trinary_state_t s = sent->symbols[i].state;
        trinary_state_t e = echoed->symbols[i].state;
        
        // XOR logic for trinary states
        if (s == STATE_MAYBE && e == STATE_MAYBE) {
            // θ ⊕ θ = 0 (verified as MAYBE)
            continue;
        } else if ((s == STATE_YES && e == STATE_NO) || 
                   (s == STATE_NO && e == STATE_YES)) {
            // Conjugate pair verified
            continue;
        } else {
            return 0;  // Invalid state
        }
    }
    
    return 1;
}

int main() {
    // Create message: YES, MAYBE, NO, MAYBE
    nsigii_symbol_t symbols[] = {
        {STATE_YES, 0.0},
        {STATE_MAYBE, 0.0},
        {STATE_NO, 0.0},
        {STATE_MAYBE, 0.0}
    };
    
    nsigii_message_t message = {
        .symbols = symbols,
        .length = 4,
        .entropy_bits = NULL
    };
    
    printf("=== NSIGII C Implementation ===\n");
    
    // Encode and measure entropy
    encode_sparse(&message);
    
    // Calculate work
    double work = calculate_work(&message, 1.25, 15.0);
    printf("Work Required: %.4f Joules\n", work);
    
    // Apply isomorphism
    nsigii_message_t echo;
    apply_isomorphism(&message, &echo);
    
    printf("Echo verified: %s\n", verify_echo(&message, &echo) ? "✓ VERIFIED" : "✗ FAILED");
    
    // Clean up
    free(message.entropy_bits);
    free(echo.symbols);
    free(echo.entropy_bits);
    
    return 0;
}